src/substrait/textplan/parser/grammar/SubstraitPlanParser.g4

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// $antlr-format alignTrailingComments on, columnLimit 130, minEmptyLines 1, maxEmptyLinesToKeep 1, reflowComments off
+// $antlr-format useTab off, allowShortRulesOnASingleLine off, allowShortBlocksOnASingleLine on, alignSemicolons ownLine
+
+parser grammar SubstraitPlanParser;
+
+options { tokenVocab = SubstraitPlanLexer; }
+
+// Follows directly after the standard #includes in h + cpp files.
+@ parser :: postinclude
+{
+#ifndef _WIN32
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+#endif
+}
+
+
+plan
+   : plan_detail* EOF
+   ;
+
+plan_detail
+   : pipelines
+   | relation
+   | schema_definition
+   | source_definition
+   | extensionspace
+   ;
+
+pipelines
+   : PIPELINES LEFTBRACE (pipeline SEMICOLON)* RIGHTBRACE
+   ;
+
+pipeline
+   : pipeline ARROW relation_ref
+   | relation_ref
+   ;
+
+// TODO -- Make the token order involving ids consistent between relations and other top-level entities.
+relation
+   : relation_type RELATION relation_ref LEFTBRACE relation_detail* RIGHTBRACE
+   ;
+
+relation_type
+   : id
+   ;
+
+relation_ref
+   : id (LEFTPAREN SCHEMA id RIGHTPAREN)?
+   ;
+
+relation_detail
+   : COMMON SEMICOLON                # relationCommon
+   | BASE_SCHEMA id SEMICOLON        # relationUsesSchema
+   | FILTER operation SEMICOLON      # relationFilter
+   | PROJECTION SEMICOLON            # relationProjection
+   | EXPRESSION operation SEMICOLON  # relationExpression
+   | ADVANCED_EXTENSION SEMICOLON    # relationAdvancedExtension
+   | source_reference SEMICOLON      # relationSourceReference
+   ;
```

Comment:
I have a table of all of the kinds of operations that can be done in relations.  The terms aren't consistently used but I've started merging the similar ones.  For instance, in two PRs you'll see BESTEFFORT and POSTJOIN qualifiers for filters which consolidate the syntax for read and join relations.  I've also decided to use FILTER instead of CONDITION so the filter clause works for filter relations as well.  The only new clause required for the existing physical relations (hash and merge join) would be a way of specifying the left keys and the right keys.  There are typically one or two new clauses required for each new relation.  (The only scary relations are the currently inaccessible write relations which would require 12 new clauses but fortunately all three share the same new clauses.)

Reply:
How do you feel about something like this (please forgive my horrible ANTLR abilities):

```
relation_key_value
   : id COLON relation_value

relation_value
   : COMMON # relationCommon (ideally this is part of relation_detail and not relation_obj but I can't ANTLR it out at the moment)
   | SCHEMA id # Schema reference
   | RELATION id # Input relation reference
   | EXPRESSION operation  # An expression
   | FUNCTION id # A function
   | LITERAL constant # A basic configurable property
   | [ relation_value* ] # Er...joined with COMMA, however you say that
   | { relation_key_value* }
   ;
relation_detail
   : relation_key_value* # I lost the SEMICOLON somewhere but we can put it back if desired
```

In other words, we end up with something very similar to JSON.  However, instead of our base types being string, number, boolean, null (all of which can be expressed with `LITERAL`) our base types also include schema, input, expression, and function.

Examples:

```
project relation project1 {
  input: relation scan
  expressions: [
    expression o_orderkey,
    expression o_custkey
  ]
}
```

```
join relation join1 {
  left: relation project1
  right: relation project2
  expression: expression equal(o_customerid, c_id)
  post_join_filter: expression equal(c_id, 7)
  type: literal "JOIN_TYPE_INNER" # Well, it's an enum, but using string for now, we can figure out enums
}

custom_extension_relation relation cer1 {
  left: relation project1
  middle: relation project2
  right: relation project3
  fizziness: literal 73.4
  approach: expression gt(inp1, 12)
}
```

(feel free to add back in semicolons or get rid of the colons or tweak however.  I'm just trying to communicate the broader idea)

Pros
 * Can represent any relation without advanced knowledge (this is the primary motivation)
 * Naturally provides for advanced_extension, etc.

Cons
 * Symbol table is more "abstract" (e.g. there won't be a "project" object in the symbol table anywhere.  However, I think something like this could still exist for the common logical types, it would just be a separate representation)
 * Serializing to protobuf will be tricky (will require reflection and use of descriptors.  I'm only "mostly sure" this is possible)
 * Slightly more verbose

Reply:
For closure, @EpsilonPrime and I spoke about this offline.  It's an interesting idea and may be something we need for extensions.  However, sticking with the current approach for the logical / popular relations makes a lot of sense.